File: 43.multiply-strings.py
#
# @lc app=leetcode id=43 lang=python3
#
# [43] Multiply Strings
#

# @lc code=start


from itertools import dropwhile, product


class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        n1, n2 = len(num1), len(num2)
        digits = [0] * (n1 + n2)

        for i, j in product(filter(lambda i: num1[i] != "0", range(n1 - 1, -1, -1)), range(n2 - 1, -1, -1)):
            digits[i + j + 1] += (ord(num1[i]) - ord("0")) * (ord(num2[j]) - ord("0"))
            digits[i + j] += digits[i + j + 1] // 10
            digits[i + j + 1] %= 10
        return "".join(map(str, dropwhile(lambda d: d == 0, digits))) or "0"


# @lc code=end

File: 43.multiply-strings.rs
/*
 * @lc app=leetcode id=43 lang=rust
 *
 * [43] Multiply Strings
 */

// @lc code=start
impl Solution {
    pub fn multiply(num1: String, num2: String) -> String {
        if num1 == "0" || num2 == "0" {
            "0".to_owned()
        } else {
            num1.bytes()
                .rev()
                .map(|b| b - b'0')
                .enumerate()
                .filter(|(_, n)| *n != 0)
                .flat_map(|(i, n1)| {
                    num2.bytes()
                        .rev()
                        .enumerate()
                        .map(move |(j, n2)| (i + j, n1 * (n2 - b'0')))
                })
                .fold(vec![0; num1.len() + num2.len()], |mut digits, (k, sum)| {
                    digits[k] += sum;
                    digits[k + 1] += digits[k] / 10;
                    digits[k] %= 10;
                    digits
                })
                .into_iter()
                .rev()
                .skip_while(|&n| n == 0)
                .map(|b| (b'0' + b) as char)
                .collect()
        }
    }
}
// @lc code=end
    